homeassistant/components/scrape/__init__.py

```diff
@@ -1 +1,141 @@
 """The scrape component."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import Any
+
+import httpx
+import voluptuous as vol
+
+from homeassistant.components.rest.data import RestData
+from homeassistant.components.sensor import (
+    CONF_STATE_CLASS,
+    DEVICE_CLASSES_SCHEMA,
+    STATE_CLASSES_SCHEMA,
+)
+from homeassistant.const import (
+    CONF_ATTRIBUTE,
+    CONF_AUTHENTICATION,
+    CONF_DEVICE_CLASS,
+    CONF_HEADERS,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_RESOURCE,
+    CONF_SCAN_INTERVAL,
+    CONF_UNIQUE_ID,
+    CONF_UNIT_OF_MEASUREMENT,
+    CONF_USERNAME,
+    CONF_VALUE_TEMPLATE,
+    CONF_VERIFY_SSL,
+    HTTP_BASIC_AUTHENTICATION,
+    HTTP_DIGEST_AUTHENTICATION,
+    Platform,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import discovery
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import CONF_INDEX, CONF_SELECT, DEFAULT_NAME, DEFAULT_VERIFY_SSL, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+SCRAPE_CONFIG = vol.Schema(
+    {
+        vol.Required(CONF_RESOURCE): cv.string,
+        vol.Required(CONF_SELECT): cv.string,
+        vol.Optional(CONF_ATTRIBUTE): cv.string,
+        vol.Optional(CONF_INDEX, default=0): cv.positive_int,
+        vol.Optional(CONF_AUTHENTICATION): vol.In(
+            [HTTP_BASIC_AUTHENTICATION, HTTP_DIGEST_AUTHENTICATION]
+        ),
+        vol.Optional(CONF_HEADERS): vol.Schema({cv.string: cv.string}),
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string,
+        vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+        vol.Optional(CONF_STATE_CLASS): STATE_CLASSES_SCHEMA,
+        vol.Optional(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
+        vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+        vol.Optional(CONF_UNIQUE_ID): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL): cv.positive_int,
+    }
+)
```

Comment:
The great thing about the Rest implementation is that you can now have multiple sensors on a single resource.
So if you want to scrape five sensors from the same page, then you don't need to load the page five times...
Could we do that here?

Note: maybe you want to keep this PR as a "everything we want to implement, but too big to review", and we can add some of the new features gradually.
For example, we can keep unique_id, device_class, state_class, etc. for a separate PR

Reply:
> Could we do that here?

Good idea. Let's do that.

> Note: maybe you want to keep this PR as a "everything we want to implement, but too big to review", and we can add some of the new features gradually. For example, we can keep unique_id, device_class, state_class, etc. for a separate PR

Sure. Sounds reasonable to split it out a bit.
Wasn't able to clean it out completely yesterday so will continue with that.

Reply:
I believe this one is sort of ready to go once the PR's you made are merged.